### DISCUSSION TITLE
Test cases for Markdown files

### DISCUSSION THREAD

**Discussion by lasyare** [mkdocs/mkdocs#3405] [General]
Given that I have a valid MkDocs project with a Markdown file, when I run the MkDocs build command, then the Markdown file should be correctly rendered into HTML.There is currently no easy way to tell whether change to the MKDocs pages will render correctly.So, how we can provide a way of testing changes before they are deployed.
for example- Test Markdown file rendering
Test Markdown extensions or plugins
Test file references
Test custom templates or themes
Test front matter or metadata

**Comment by pawamoy** [comment]
I suppose you can write automated tests in Bash or Python, that build the site locally, in strict mode, and then inspect the resulting HTML files?

**Reply by lasyare** [reply]
Automation would be my next idea but first I am planning to write down the scenario so just needed some scenarios like mentioned above in the description

**Reply by lasyare** [reply]
perhaps something like pre merge and post merge scenarios , while merging any test cases can be included

**Comment by pawamoy** [comment]
I think... you are looking for `mkdocs serve`? This command allows you to browse your site locally.

**Comment by lasyare** [comment]
a kind of mk docs serve as well, just the test cases or scenarios i am looking in.

**Reply by pawamoy** [reply]
I'm still not sure to understand what you're trying to do.

> how we can provide a way of testing changes before they are deployed

IMO that's `mkdocs serve` and the `--strict` option.

> Test Markdown file rendering
Test Markdown extensions or plugins
Test file references
Test custom templates or themes
Test front matter or metadata

Most plugins and extensions will report errors. MkDocs itself will report invalid links. For the rest you simply serve and check out each page manually.

If I needed formalism, I'd write: "Serve docs locally and verify that each page is correctly rendered". Maybe add a few things like "Assert that links work and point to the right resources". Maybe more, but that depends on your site contents :shrug:

**Comment by kamilkrzyskow** [comment]
I'm not entirely sure what you're after either, when you ask for Markdown file test cases, but in our docs we do test Markdown files before build.

https://github.com/Gothic-Modding-Community/gmc/blob/main/tests/test_indentation.py

In general we make sure that admonitions have 4 spaces of indentation to render properly and a few other things too. 
Those tests are rather heavily opinionated, like the empty line before a listing, because as it turned out that too isn't completely right and can be solved with a 4 space indentation as well. Only learned about it recently from https://github.com/squidfunk/mkdocs-material/discussions/6048

I think the 4 space indentation test would be also redundant if we implemented a special IDE file that would enforce 4 spaces indentation everywhere, but I never came around to creating it 🤷‍♂️

IMO you should think about things that might have broke in the past and made your MkDocs build invalid. Only this warrants creating tests. 
So do you have any case like that?

**Comment by lasyare** [comment]
yes i do have some but let me check out what is provided int he link you have provided so that I can prepare altogether maximum coverage of test scenarios . Thank you
